beacon/goclient/attest.go

```diff
@@ -27,22 +27,106 @@ func (gc *GoClient) AttesterDuties(ctx context.Context, epoch phase0.Epoch, vali
+// GetAttestationData returns attestation data for a given slot (which is same for all 64 committeeIndex
+// values that identify Ethereum committees chosen to attest on this slot).
+// Note, committeeIndex is an optional parameter that will be used to set AttestationData.Index
+// in the resulting data returned from this function.
+// Note, result returned is meant to be read-only, it's not safe to modify it (because it will be
+// accessed by multiple concurrent readers).
+func (gc *GoClient) GetAttestationData(slot phase0.Slot, committeeIndex phase0.CommitteeIndex) (
+	result *phase0.AttestationData,
+	version spec.DataVersion,
+	err error,
+) {
+	// Final processing for result returned.
+	defer func() {
+		if err != nil {
+			// Nothing to process, just propagate error.
+			return
+		}
+
+		// Assign committeeIndex passed to GetAttestationData call, the rest of attestation data stays
+		// unchanged.
+		// Note, we cannot return result object directly here modifying its Index value because it
+		// would be unsynchronised concurrent write (since it's cached for other concurrent readers
+		// to access). Hence, we return shallow copy here. We don't need to return deep copy because
+		// the callers of GetAttestationData will only use this data to read it (they won't update it).
+		result = &phase0.AttestationData{
+			Slot:            result.Slot,
+			Index:           committeeIndex,
+			BeaconBlockRoot: result.BeaconBlockRoot,
+			Source:          result.Source,
+			Target:          result.Target,
+		}
+	}()
+
+	// Check cache.
+	cachedResult, ok := gc.attestationDataCache.Get(slot)
+	if ok {
+		return cachedResult, spec.DataVersionPhase0, nil
+	}
+
+	// Have to make beacon node request and cache the result.
 	attDataReqStart := time.Now()
-	resp, err := gc.client.AttestationData(gc.ctx, &api.AttestationDataOpts{
-		Slot:           slot,
-		CommitteeIndex: committeeIndex,
-	})
+	result, err = func() (*phase0.AttestationData, error) {
+		// Requests with the same slot number must lock on the same mutex.
+		reqMu := &gc.attestationReqMuPool[int64(slot)%int64(len(gc.attestationReqMuPool))]
+		reqMu.Lock()
+		defer reqMu.Unlock()
+
+		// Prevent making more than 1 beacon node requests in case somebody has already made this
+		// request concurrently and succeeded.
+		cachedResult, ok := gc.attestationDataCache.Get(slot)
+		if ok {
+			return cachedResult, nil
+		}
+
+		resp, err := gc.client.AttestationData(gc.ctx, &api.AttestationDataOpts{
+			Slot: slot,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get attestation data: %w", err)
+		}
+		if resp == nil {
+			return nil, fmt.Errorf("attestation data response is nil")
+		}
+
+		gc.attestationDataCache.Set(slot, resp.Data)
+		gc.recentAttestationSlot.Store(uint64(slot))
+
+		return resp.Data, nil
+	}()
+	metricsAttesterDataRequest.Observe(time.Since(attDataReqStart).Seconds())
```

Comment:
What do you think about moving this inside the function that gets attestation data?

Reply:
Moved like you suggested.